src/arewefluentyet/source.py

```diff
@@ -0,0 +1,77 @@
+import subprocess
+import os
+from datetime import date
+
+
+def parse_date(input):
+    return date(*(int(s) for s in input.split("-")))
+
+
+class Source:
+    path = None
+    current_revision = None
+
+    def __init__(self, path):
+        self.path = path
+
+    def get_current_revision(self):
+        if not self.current_revision:
+            result = subprocess.run([
+                "hg", "id", "--cwd", self.path, "-T{id}"
+            ], check=True, capture_output=True, text=True)
+            self.current_revision = result.stdout
+        return self.current_revision
+
+    def pick_next_revision(self, next_date):
+        result = subprocess.run([
+            "hg", "log",
+            "--cwd", self.path,
+            "-l", "1",
+            "-T", "{node}",
+            "-r",
+            f"reverse(pushhead() and pushdate('< {next_date}') and ::central)"
+        ], check=True, capture_output=True, text=True)
+        return result.stdout
+
+    def get_revision_date(self, rev):
+        result = subprocess.run([
+            "hg", "id", "--cwd", self.path,
+            "-r", rev, "-T", "{pushdate|shortdate}"
+        ], check=True, capture_output=True, text=True)
+        return parse_date(result.stdout)
+
+    def switch_to_revision(self, rev):
+        if rev == self.current_revision:
+            return
+
+        subprocess.run([
+            "hg", "status", "--cwd", self.path, "-mard"
+        ], check=True)
+
+        subprocess.run([
+            "hg", "update", "--cwd", self.path, "-c", "-r", rev
+        ], check=True)
+        self.current_revision = rev
+
+    def get_bookmark_parent(self, bookmark):
+        result = subprocess.run([
+            "hg", "log",
+            "--cwd", self.path,
+            "-r", bookmark,
+            "-T", "{p1.node}",
+        ], check=True, capture_output=True, text=True)
+
+        return result.stdout
+
+    def rebase_bookmark(self, revision, bookmark):
+        parent = self.get_bookmark_parent(bookmark)
+        if parent != revision:
+            subprocess.run([
+                "hg", "rebase", "--cwd", self.path,
+                "-s", bookmark, "-d", revision
```

Comment:
`-b` might be better than `-s` in this case. Probably nothing new in this patch, just a ride-along.